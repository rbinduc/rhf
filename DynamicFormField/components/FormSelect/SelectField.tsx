import React, {useRef, useState, CSSProperties, SyntheticEvent} from 'react';
import {createStyles, makeStyles, MenuItem, Select, SelectProps} from '@material-ui/core';
import {ArrowExpand} from '../../../../../icons/global/common/ArrowExpand';
import {useTheme} from 'emotion-theming';
import {ISelectElementPaper, ITenantTheme} from '../../../../../../models';
import {ControllerRenderProps, FieldValues} from 'react-hook-form';
import {baseImageURL} from '../../../../../../helpers/HelperUtils';
import styled from '@emotion/styled';
import {Typography} from '../../../../../atoms/Typography';
import {Keyboard} from '@ionic-native/keyboard';

export interface ISelectOption {
  label: any;
  value: any;
  hasFlag?: boolean;
  max?: number;
  groupBy?: string;
}

export interface ISelectFieldProps extends SelectProps {
  options?: ISelectOption[];
  ariaLabel?: string;
  field?: ControllerRenderProps<FieldValues, any>;
  isDirty?: boolean;
}

interface ISelectFieldCSSProps extends ISelectElementPaper {
  backgroundColor: CSSProperties['backgroundColor'];
}

const useStyles = makeStyles(() =>
  createStyles({
    root: {
      position: 'relative',
    },
    list: {
      padding: 0,
      border: ({border}: ISelectFieldCSSProps) => border,
      boxSizing: 'border-box',
      borderRadius: ({borderRadius}: ISelectFieldCSSProps) => borderRadius,
    },
    rootMenuItem: {
      '&.MuiListItem-root.Mui-selected, .MuiListItem-root.Mui-selected': {
        backgroundColor: ({selectedColor}: ISelectFieldCSSProps) => selectedColor,
        '&:first-of-type': {
          borderRadius: '0.313rem 0.313rem 0 0',
        },
        '&:last-of-type': {
          borderBottom: 'unset',
          borderRadius: '0 0 0.313rem 0.313rem',
        },
        '&.MuiButtonBase-root:active': {
          backgroundColor: ({backgroundColor}: ISelectFieldCSSProps) => backgroundColor,
        },
      },

      borderBottom: ({borderBottom}: ISelectFieldCSSProps) => borderBottom,
    },
    rootSelect: {
      display: 'flex',
      justifyContent: 'flex-start',
      '&.MuiSelect-select:focus': {
        backgroundColor: 'unset',
      },
    },
  })
);

const SelectField = ({
  id,
  label,
  labelId,
  placeholder,
  ariaLabel,
  options,
  required,
  disabled,
  field,
  isDirty,
  ...rest
}: ISelectFieldProps) => {
  const selectFieldref = useRef<HTMLDivElement>(null);
  const theme: ITenantTheme = useTheme();
  const {border, borderRadius, borderBottom, selectedColor}: ISelectElementPaper = theme?.designSystem?.atoms?.formSelectElementPaper ?? {};
  const {backgroundColor} = theme.designSystem.atoms.button.hover;
  const [open, setOpen] = useState(false);

  const handleClose = () => {
    !disabled && setOpen(false);
  };

  const handleOpen = () => {
    if (!disabled) {
      if (selectFieldref?.current && typeof selectFieldref.current.scrollIntoView === 'function') {
        selectFieldref.current.scrollIntoView({block: 'center'});
      }
      if (Keyboard.isVisible) {
        setTimeout(() => {
          setOpen(true);
        }, 600);
      } else {
        setOpen(true);
      }
    }
  };

  const classes = useStyles({border, borderRadius, borderBottom, selectedColor, backgroundColor});
  return !!options?.length ? (
    <div ref={selectFieldref} className={classes.root}>
      <Select
        {...rest}
        {...field}
        labelId={labelId}
        placeholder={placeholder}
        IconComponent={!disabled ? ArrowExpand : () => <span />}
        classes={{root: classes.rootSelect}}
        className={isDirty ? 'dirty' : ''}
        variant="outlined"
        fullWidth
        disabled={disabled}
        open={open}
        onClose={handleClose}
        onOpen={handleOpen}
        inputProps={{
          id,
          required,
          disabled,
          readOnly: disabled,
          'aria-label': ariaLabel || (label && typeof label === 'string' ? label : placeholder),
          'aria-hidden': disabled,
          ...rest?.inputProps,
        }}
        MenuProps={{
          anchorOrigin: {
            vertical: 'bottom',
            horizontal: 'left',
          },
          elevation: 0,
          getContentAnchorEl: null,
          transformOrigin: {
            vertical: 'top',
            horizontal: 'left',
          },
          classes: {list: classes.list},
          transitionDuration: 800,
        }}
      >
        {options?.map(({label: optionLabel, value: optionValue, hasFlag}: ISelectOption) => (
          <MenuItem disabled={!open} button key={optionValue} selected value={optionValue} classes={{root: classes.rootMenuItem}}>
            {hasFlag && (
              <FlagStyled>
                <ImgStyled
                  src={`${baseImageURL}/countries/${optionLabel}.svg`}
                  onError={(event: SyntheticEvent<HTMLImageElement, Event>) => {
                    if ((event?.target as HTMLImageElement)?.src) {
                      (event.target as HTMLImageElement).src = `${baseImageURL}/countries/default.svg`;
                    }
                  }}
                />
              </FlagStyled>
            )}
            <Label variant="h4">{optionLabel}</Label>
          </MenuItem>
        ))}
      </Select>
      {!field?.value && <Placeholder onClick={handleOpen}>{placeholder}</Placeholder>}
    </div>
  ) : null;
};

const FlagStyled = styled.div`
  padding-right: 1.875rem;
  width: 1.5rem;
  height: 1.5rem;
  box-sizing: border-box;
`;

const ImgStyled = styled.img`
  width: 1.25rem;
  border-radius: 50%;
`;

const Label = styled(Typography)`
  margin-top: 0.125rem;
  height: auto;
  overflow: hidden;
  min-height: 1.1876em;
  white-space: nowrap;
  text-overflow: ellipsis;
`;

const Placeholder = styled(Typography)`
  position: absolute;
  left: 0;
  right: 0;
  opacity: 0.3;
  padding: 1rem;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
`;

export default SelectField;
