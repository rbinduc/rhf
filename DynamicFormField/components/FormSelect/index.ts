export {default} from './FormSelect.component';
export type {IFormSelectProps} from './FormSelect.component'
export type {ISelectOption} from './SelectField'