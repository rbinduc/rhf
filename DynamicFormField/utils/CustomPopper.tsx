import React from 'react';
import {Popper, PopperProps} from '@material-ui/core';

export const CustomPopper = ({placement = 'bottom', ...rest}: PopperProps) => <Popper {...rest} placement={placement} />;
