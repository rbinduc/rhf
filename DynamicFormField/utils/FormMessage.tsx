import styled from '@emotion/styled';
import {FormHelperText} from '@material-ui/core';
import React from 'react';
import {IconNames} from '../../../../icons/IconNameConstants';
import {TenantVector} from '../../../../icons/TenantVector/TenantVector.component';

interface IMessageProps {
  message?: React.ReactNode;
  isBold?: boolean;
  isLarge?: boolean;
}

export interface IFormMessageProps extends IMessageProps {
  id?: string;
  Icon?: IconNames;
  component?: React.ElementType;
  margin?: 'dense';
  isError?: boolean;
  hideIcon?: boolean;
}

const FormMessage = ({id = '', Icon, message, component, margin, isBold, isLarge, hideIcon, isError = true}: IFormMessageProps) =>
  message ? (
    <FormHelperText id={id} error={!!isError} component={component || 'div'} margin={margin}>
      <FormMessageWrapper>
        {!hideIcon && (
          <span id={id + '_icon'}>
            <TenantVector tenantVectorName={Icon || (isError ? IconNames.FieldError : IconNames.Info)} />
          </span>
        )}
        <Message id={id + '_message'} isBold={!!isBold} isLarge={!!isLarge}>
          {message}
        </Message>
      </FormMessageWrapper>
    </FormHelperText>
  ) : null;

const FormMessageWrapper = styled.span`
  display: flex;
  flex-direction: row;
  & svg {
    margin-top: 0.2rem;
    margin-right: 0.5rem;
  }
  word-break: break-word;
`;

const Message = styled.span<IMessageProps>`
  font-weight: ${({isBold}) => (isBold ? 'bold' : 400)};
  font-size: ${({isLarge}) => (isLarge ? '1rem' : '0.875rem')};
`;

export default FormMessage;
