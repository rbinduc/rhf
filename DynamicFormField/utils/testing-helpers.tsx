import React from 'react';
import {render} from '@testing-library/react';
import {FormProvider} from 'react-hook-form';
import useFormHook from './hooks/useFormHook';
import {ThemeProvider} from 'emotion-theming';
import {themeMock} from '../../../../themes/theme-mock';

interface IDefaultValue {
  [key: string]: any;
}

interface RenderOptions {
  defaultValues?: IDefaultValue;
  toPassBack?: any[];
}

/**
 * Testing Library utility function to wrap tested component in React Hook Form
 * @param {ReactElement} ui A React component
 * @param objectParameters
 * @param {Object} objectParameters.defaultValues Initial form values to pass into
 * React Hook Form, which you can then assert against
 * @param {Array} objectParameters.toPassBack React Hook Form method names which we'd
 * like to pass back and use in tests.  A primary use case is sending back 'setError',
 * so we can manually setErrors on React Hook Form components and test error handling
 */
export function renderWithDyanmicForm(
  ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
  {defaultValues = {}, toPassBack = []}: RenderOptions = {}
) {
  let dynamicFormMethods = {} as any;

  const Wrapper = ({children}: {children: React.ReactNode}) => {
    const {methods}: any = useFormHook({defaultValues});
    for (let dynamicFormItem of toPassBack) {
      dynamicFormMethods[dynamicFormItem] = methods[dynamicFormItem];
    }

    return (
      <ThemeProvider theme={themeMock}>
        <FormProvider {...methods}>{children}</FormProvider>
      </ThemeProvider>
    );
  };

  return {
    ...render(ui, {wrapper: Wrapper as React.ComponentType<{}>}),
    dynamicFormMethods,
  };
}

/**
 * Higher order helper function which wraps a component w/ React Hook Form
 * @param {React Component} WrappedComponent to pass into
 * @param {*} restProps any other remaining props
 * @returns {React Component}
 */
export function withDynamicForm(WrappedComponent: React.ComponentType<any>, restProps: object, defaultValues: IDefaultValue) {
  return () => {
    const methods = useFormHook({defaultValues});

    return (
      <ThemeProvider theme={themeMock}>
        <WrappedComponent {...restProps} {...methods} />
      </ThemeProvider>
    );
  };
}
