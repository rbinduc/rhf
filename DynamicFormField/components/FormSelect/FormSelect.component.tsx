import React from 'react';
import styled from '@emotion/styled';
import {FormControl, InputLabel} from '@material-ui/core';
import {Controller, useFormContext} from 'react-hook-form';
import SelectField, {ISelectFieldProps} from './SelectField';
import {IValidations} from '../../../DynamicForm';
import {getLabel} from '../../DynamicFormField';
import FormMessage from '../../utils/FormMessage';

export type IFormSelectProps = ISelectFieldProps & {
  isDisabled?: boolean;
  isRequired?: boolean;
  labelId?: string;
  validations?: IValidations;
  hideAsterisk?: boolean;
};

const FormSelect: React.FC<IFormSelectProps> = ({
  name,
  label,
  defaultValue,
  isRequired,
  isDisabled,
  hideAsterisk,
  options,
  labelId,
  validations,
  ...other
}) => {
  const {
    control,
    formState: {errors, dirtyFields},
  } = useFormContext();
  const errorMessage = errors[name!]?.message;
  const isDirty = !!dirtyFields[name!];
  return (
    <FormControl variant="outlined" required={isRequired} fullWidth disabled={isDisabled} error={Boolean(errorMessage)}>
      {label && (
        <InputLabelStyled shrink={false} htmlFor={name} className={errorMessage && 'error'} id={labelId}>
          {getLabel(label, !!isRequired, hideAsterisk)}
        </InputLabelStyled>
      )}
      <Controller
        name={name as any}
        control={control}
        defaultValue={defaultValue}
        rules={validations}
        render={({field}) => (
          <SelectField
            label={label}
            labelId={labelId}
            options={options!}
            required={!!isRequired}
            disabled={!!isDisabled}
            field={field}
            isDirty={isDirty}
            {...other}
          />
        )}
      />
      <FormMessage message={errorMessage} margin="dense" />
    </FormControl>
  );
};

const InputLabelStyled = styled(InputLabel)`
  .MuiFormLabel-asterisk {
    display: none;
  }
`;

export default FormSelect;
