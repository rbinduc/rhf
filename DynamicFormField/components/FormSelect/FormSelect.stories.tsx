import React from 'react';
import {storiesOf} from '@storybook/react';
import FormSelect from '.';
import {IFormSelectProps} from './FormSelect.component';
import {ThemeProvider} from 'emotion-theming';
import {themeMock} from '../../../../../themes/theme-mock';
import {FormProvider} from 'react-hook-form';
import useFormHook from '../../utils/hooks/useFormHook';

storiesOf('Organisms/RHF/DynamicForm/FormSelect', module)
  .addDecorator((getStory) => <ThemeProvider theme={themeMock}>{getStory()}</ThemeProvider>)

  .add('Langauge', () => {
    const defaultProps: IFormSelectProps = {
      name: 'Language Selection',
      label: 'Language',
      defaultValue: 'en-US',
      options: [
        {
          label: 'English',
          value: 'en-US',
        },
        {
          label: 'Espa√±ol',
          value: 'es-US',
        },
      ],
      isRequired: true,
      id: 'language',
    };
    const {methods}: any = useFormHook({});
    return (
      <FormProvider {...methods}>
        <FormSelect {...defaultProps} />
      </FormProvider>
    );
  })

  .add('Country', () => {
    const defaultProps: IFormSelectProps = {
      name: 'Country Selection',
      label: 'Country / region',
      defaultValue: 101,
      options: [
        {
          label: 'United States of America',
          value: 101,
          hasFlag: true,
        },
        {
          label: 'Canada',
          value: 100,
          hasFlag: true,
        },
      ],
      isRequired: true,
      id: 'country',
    };
    const {methods}: any = useFormHook({});
    return (
      <FormProvider {...methods}>
        <FormSelect {...defaultProps} />
      </FormProvider>
    );
  });
