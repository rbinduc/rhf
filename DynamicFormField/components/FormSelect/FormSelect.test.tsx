import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import FormSelect, {IFormSelectProps} from '.';
import {renderWithDyanmicForm} from '../../utils/testing-helpers';

import {screen} from '@testing-library/react';
const defaultProps: IFormSelectProps = {
  name: 'Country Selection',
  label: 'Country / region',
  defaultValue: 101,
  options: [
    {
      label: 'United States of America',
      value: 101,
      hasFlag: true,
    },
    {
      label: 'Canada',
      value: 100,
      hasFlag: true,
    },
  ],
  isRequired: true,
  id: 'country',
};

jest.mock('@ionic-native/keyboard', () => ({
  hide: jest.fn(),
}));

jest.mock('../../../../../../automation/PageLocators', () => ({
  locators: {
    jobPageLocators: {
      aFiltersIcon: 'filtersIcon',
    },
    commonLocators: {
      aHeaderText: 'headerText',
    },
    accountSettings: {
      aChangeCountryHeader: 'changeCountryHeader',
    },
  },
}));

describe('FormSelect', () => {
  it('snapshot match', () => {
    const {container} = renderWithDyanmicForm(<FormSelect {...defaultProps} />, {});
    expect(container).toBeDefined();
    expect(container.firstChild).toMatchSnapshot();

    expect(screen).toMatchSnapshot();
  });
});
