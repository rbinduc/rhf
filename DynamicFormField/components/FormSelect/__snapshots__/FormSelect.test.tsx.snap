// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSelect snapshot match 1`] = `
<div
  class="MuiFormControl-root MuiFormControl-fullWidth"
>
  <label
    class="MuiFormLabel-root MuiInputLabel-root css-ftyd21-InputLabelStyled e1xmma6j0 MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-filled Mui-required Mui-required"
    data-shrink="false"
    for="Country Selection"
  >
    Country / region*
    <span
      aria-hidden="true"
      class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
    >
       
      *
    </span>
  </label>
  <div
    class="makeStyles-root-1"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
    >
      <div
        aria-haspopup="listbox"
        aria-label="Country / region"
        aria-labelledby="mui-component-select-Country Selection"
        class="MuiSelect-root makeStyles-rootSelect-4 MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
        id="mui-component-select-Country Selection"
        role="button"
        tabindex="0"
      >
        <div
          class="css-wt89bw-FlagStyled e1rb4ngw0"
        >
          <img
            class="css-yqvute-ImgStyled e1rb4ngw1"
            src="/images/countries/United States of America.svg"
          />
        </div>
        <h4
          class="MuiTypography-root css-164bkhw-Label e1rb4ngw2 MuiTypography-h4"
        >
          United States of America
        </h4>
      </div>
      <input
        aria-hidden="false"
        class="MuiSelect-nativeInput"
        id="country"
        name="Country Selection"
        required=""
        tabindex="-1"
        value="101"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
        focusable="false"
        style="width: 18px; height: 24px;"
        viewBox="0 0 18 18"
      >
        <path
          d="M5.23828 7.48828H12.7617L9.70711 10.5429C9.31658 10.9334 8.68342 10.9334 8.29289 10.5429L5.23828 7.48828Z"
          fill="#273446"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
        style="padding-left: 8px;"
      >
        <legend
          class="PrivateNotchedOutline-legend-8"
          style="width: 0.01px;"
        >
          <span>
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`FormSelect snapshot match 2`] = `
Object {
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;
